src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -687,4 +753,102 @@ public static string GetTempPath(string workingDir)
             return Environment.GetEnvironmentVariable("SYSTEMROOT");
         }
     }
+
+    internal class FileMutex : IDisposable 
+    {
+        public readonly bool OwnsMutex;
+        public readonly FileStream Stream;
+        public readonly string FilePath;
+
+        public bool Locked { get; private set; }
+
+        public FileMutex(string name) 
+        {
+            FilePath = Path.Combine(Path.GetTempPath(), name);
```

Comment:
The API `Path.GetTempPath` can't be safely used here. The value `%TEMP%` is not constant through the life of an entire build operation. It's awful but build environments will change the value around during a build. We've learned this lesson the hard way. I think we need to pick a more stable place to put any temporary files when both processes have to agree on the location.

Reply:
Hmm I'm not sure of a good place. @nguerrera is most likely to have encountered this in the CLI space and may have a suggestion.

Reply:
I'll ditch this API but I just noticed the file uses it elsewhere... should I fix that too?

Reply:
We write everything temporary to somewhere under the project's `$(IntermediateOutputPath)`.

Reply:
I realize that probably isn't helpful here.

Reply:
I went with a 3-level fallback:
%localappdata%\.roslyn\x
%appdata%\.roslyn\x
%temp%\.roslyn\x
I feel like that should work out okay?

Reply:
I had to refresh my memory a bit of why we had so much problem with `Path.GetTempPath` in the past. Here is the PR where I put down a fairly detailed accounting of the problem:

https://github.com/dotnet/roslyn/pull/14417

The core of the issue though is that `Path.GetTempPath` has two interesting pieces of state that can differ between build requests:

1. `%TEMP%` and `%TMP%`: evil, but very real build systems, decide to change these values between build invocations. 
1. The working directory is an input to `Path.GetTempPath` when 1 above is a relative path. The working directory for MSBuild is different than the working directory for the invoked csc / vbc executables. Hence when doing compiler server communication we had to recreate `Path.GetTempPath` to take the working directory as an input parameter so we could emulate `Path.GetTempPath` without having to start a process to do so. 

In thinking through this I don't think we can have a perfect solution here. The presence of 1 above pretty much eliminates that. But I do think we need to account for 2. If we don't then any use of relative paths in `%TEMP%` and `%TMP%` will result in lots of compiler servers getting spawned. 

I think the right approach then is to have both sides use `BuildServerConnection.GetTempPath` for calculating the temporary path while using the directory containing the compiler server as the input parameter. That will give us consistency across the process boundaries. 

Or alternatively if this is for Unix only maybe we just write to `/tmp` directly?

Reply:
@agocke, @chsienki FYI

Reply:
I'm not sure if Mono-on-Windows-but-not-Cygwin is a supported scenario, i.e. no /tmp. I'll find out. If it's not, we could just write to /tmp but I think in my testing that would prevent two separate users from both using roslyn's compiler server, which seems less than ideal.